homeassistant/components/xiaomi_miio/humidifier.py

```diff
@@ -0,0 +1,660 @@
+    SERVICE_HUMIDIFIER_SET_MOTOR_SPEED: {
+        "method": "async_set_motor_speed",
+        "schema": SERVICE_SCHEMA_MOTOR_SPEED,
+    },
+}
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Fan from a config entry."""
+    entities = []
+
+    if config_entry.data[CONF_FLOW_TYPE] == CONF_DEVICE:
+        if DATA_KEY not in hass.data:
+            hass.data[DATA_KEY] = {}
+
+        host = config_entry.data[CONF_HOST]
+        token = config_entry.data[CONF_TOKEN]
+        name = config_entry.title
+        model = config_entry.data[CONF_MODEL]
+        unique_id = config_entry.unique_id
+
+        _LOGGER.debug("Initializing with host %s (token %s...)", host, token[:5])
+
+        if model in MODELS_HUMIDIFIER_MIOT:
+            air_humidifier = AirHumidifierMiot(host, token)
+            entity = XiaomiAirHumidifierMiot(
+                name, air_humidifier, config_entry, unique_id
+            )
+        elif model in MODELS_HUMIDIFIER:
+            air_humidifier = AirHumidifier(host, token, model=model)
+            entity = XiaomiAirHumidifier(name, air_humidifier, config_entry, unique_id)
+        else:
+            _LOGGER.error(
+                "Unsupported device found! Please create an issue at "
+                "https://github.com/syssi/xiaomi_airpurifier/issues "
+                "and provide the following data: %s",
+                model,
+            )
+            return
+
+        hass.data[DATA_KEY][host] = entity
+        entities.append(entity)
+
+        async def async_service_handler(service):
+            """Map services to methods on XiaomiAirPurifier."""
+            method = SERVICE_TO_METHOD[service.service]
+            params = {
+                key: value
+                for key, value in service.data.items()
+                if key != ATTR_ENTITY_ID
+            }
+            entity_ids = service.data.get(ATTR_ENTITY_ID)
+            if entity_ids:
+                entities = [
+                    entity
+                    for entity in hass.data[DATA_KEY].values()
+                    if entity.entity_id in entity_ids
+                ]
+            else:
+                entities = hass.data[DATA_KEY].values()
+
+            update_tasks = []
+
+            for entity in entities:
+                entity_method = getattr(entity, method["method"], None)
+                if not entity_method:
+                    continue
+                await entity_method(**params)
+                update_tasks.append(
+                    hass.async_create_task(entity.async_update_ha_state(True))
+                )
+
+            if update_tasks:
+                await asyncio.wait(update_tasks)
+
+        for air_purifier_service in SERVICE_TO_METHOD:
+            schema = SERVICE_TO_METHOD[air_purifier_service].get(
+                "schema", HUMIDIFIER_SERVICE_SCHEMA
+            )
+            hass.services.async_register(
```

Comment:
Please use our entity service helper instead.

https://developers.home-assistant.io/docs/dev_101_services#entity-services

Note that the service schema should just be a dict with voluptuous validators.